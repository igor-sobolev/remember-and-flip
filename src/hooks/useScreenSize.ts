import { useLayoutEffect, useState, RefObject } from 'react';

export function useScreenSize(container?: RefObject<HTMLDivElement>): number[] {
  const [size, setSize] = useState([0, 0]);
  useLayoutEffect(() => {
    function updateSize() {
      const width = container?.current?.clientWidth ? container.current?.clientWidth : window.innerWidth;
      const height = container?.current?.clientHeight ? container.current?.clientWidth : window.innerHeight;
      setSize([width, height]);
    }
    window.addEventListener('resize', updateSize);
    updateSize();
    return () => window.removeEventListener('resize', updateSize);
  }, [container]);
  return size;
}
