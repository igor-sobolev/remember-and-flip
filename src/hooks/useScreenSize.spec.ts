import React, { RefObject } from 'react';
import { renderHook, RenderHookResult } from '@testing-library/react-hooks';

import { useScreenSize } from './useScreenSize';

const WIDTH = 1001;
const HEIGHT = 100;

describe('Given useScreenSize hook', () => {
  let windowSpy: jest.SpyInstance;
  let rendered: RenderHookResult<RefObject<HTMLDivElement>, number[]>;

  beforeAll(() => {
    windowSpy = jest.spyOn(global, 'window', 'get');
  });

  afterAll(() => {
    windowSpy.mockRestore();
  });

  beforeEach(() => {
    windowSpy.mockImplementation(() => ({
      innerWidth: WIDTH,
      innerHeight: HEIGHT,
      addEventListener: jest.fn(),
      removeEventListener: jest.fn()
    }));
    rendered = renderHook(() => useScreenSize());
  });

  it('should match the snapshot', () => {
    expect(rendered).toMatchSnapshot();
  });

  it('should return correct result', () => {
    expect(rendered.result.current).toStrictEqual([WIDTH, HEIGHT]);
  });

  describe('when passed a container', () => {
    const NEW_WIDTH = 125;
    const NEW_HEIGHT = 120;

    beforeAll(() => {
      jest.spyOn(React, 'createRef').mockReturnValue({
        current: {
          clientWidth: NEW_WIDTH,
          clientHeight: NEW_HEIGHT
        }
      });
    });

    beforeEach(() => {
      const containerRef: React.RefObject<HTMLDivElement> = React.createRef();

      rendered = renderHook(() => useScreenSize(containerRef));
    });

    it('should match the snapshot', () => {
      expect(rendered).toMatchSnapshot();
    });

    it('should return correct result', () => {
      expect(rendered.result.current).toStrictEqual([NEW_WIDTH, NEW_HEIGHT]);
    });
  });
});
