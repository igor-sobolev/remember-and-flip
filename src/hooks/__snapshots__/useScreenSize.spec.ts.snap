// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given useScreenSize hook should match the snapshot 1`] = `
Object {
  "rerender": [Function],
  "result": Object {
    "current": Array [
      1001,
      100,
    ],
    "error": undefined,
  },
  "unmount": [Function],
  "wait": [Function],
  "waitFor": [Function],
  "waitForNextUpdate": [Function],
  "waitForValueToChange": [Function],
}
`;

exports[`Given useScreenSize hook when passed a container should match the snapshot 1`] = `
Object {
  "rerender": [Function],
  "result": Object {
    "current": Array [
      125,
      120,
    ],
    "error": undefined,
  },
  "unmount": [Function],
  "wait": [Function],
  "waitFor": [Function],
  "waitForNextUpdate": [Function],
  "waitForValueToChange": [Function],
}
`;
